design/orchestrators.md

```diff
@@ -0,0 +1,231 @@
+aren't running (in terms of observed state) over tasks that are currently
+running. Note that scale-down decisions are made by the orchestrator, and don't
+quite match the state the scheduler would arrive at when scaling up. This is an
+area for future improvement; see https://github.com/docker/swarmkit/issues/2320
+for more details.
+
+In both of these cases, and also in the case where the number of replicas is
+already correct, the orchestrator calls the update supervisor to ensure that the
+existing tasks (or tasks being kept, in the case of a scale-down) are
+up-to-date. The update supervisor does the heavy lifting involved in rolling
+updates and automatic rollbacks, but this is all abstracted from the
+orchestrator.
+
+## Global orchestrator
+
+The global orchestrator works similarly to the replicated orchestrator, but
+tries to maintain one task per active node meeting the constraints, instead of a
+specific total number of replicas. It ignores services that aren't global
+services and tasks that aren't associated with global services.
+
+The global orchestrator responds to task events in much the same way that the
+replicated orchestrator does. If a task fails, the global orchestrator will
+indicate to the restart supervisor that a restart may be needed.
+
+When a service is created, updated, or deleted, this triggers a reconciliation.
+The orchestrator has to check whether each node meets the constraints for the
+service, and create or update tasks on that node if it does. The tasks are
+created with a specific node ID pre-filled. They pass through the scheduler so
+that the scheduler can wait for the nodes to have sufficient resources before
+moving the desired state to `Assigned`, but the scheduler does not make the
+actual scheduling decision.
+
+The global orchestrator also responds to node events. These trigger
+reconciliations much like service events do. A new node might mean creating a
+task from each service on that node, and a deleted node would mean deleting any
+global service tasks from that node. When a node gets drained, the global
+orchestrator shuts down any global service tasks running on that node. It also
+does this when a node goes down, which avoids stuck rolling updates that would
+otherwise want to update the task on the unavailable node before proceeding.
+
+Like the replicated orchestrator, the global orchestrator uses the update
+supervisor to implement rolling updates and automatic rollbacks. Instead of
+passing tasks to the update supervisor by slot, it groups them by node. This
+means rolling updates will go node-by-node instead of slot-by-slot.
+
+## Restart supervisor
+
+The restart supervisor manages the process of shutting down a task, and possibly
+starting a replacement task. Its entry point is a `Restart` method which is
+called inside a store write transaction. It atomically changes the state of the
+old task to `Shutdown`, and, if it's appropriate to start a replacement task,
```

Comment:
Should we enumerate some cases where it may not be appropriate to start a replacement task?  For instance if the restart policy says no?  (checking for draining, pause, or a down node I guess happens in the orchestrator?)  Are there any other cases where it'd be inappropriate to restart?